great_expectations/metrics/metric.py

```diff
@@ -20,34 +21,62 @@ def __init__(self, class_name: str, mixin_superclass_name: str) -> None:
         )
 
 
+class MissingAttributeError(AttributeError):
+    def __init__(self, class_name: str, attribute_name: str) -> None:
+        super().__init__(f"`{class_name}` must define `{attribute_name}` attribute.")
+
+
 @dataclass_transform()
 class MetaMetric(ModelMetaclass):
-    def __new__(cls, name, bases, attrs):
-        # ensure a single Domain mixin is defined
-        if name != "Metric" and (
-            len(bases) != ALLOWABLE_METRIC_MIXINS + 1
-            or not any(issubclass(base_type, Domain) for base_type in bases)
-        ):
-            raise MixinTypeError(name, "Domain")
-        return super().__new__(cls, name, bases, attrs)
+    """Metaclass for Metric classes that maintains a registry of all concrete Metric types."""
 
+    _registry: dict[str, type["Metric"]] = {}
 
-class Metric(BaseModel, metaclass=MetaMetric):
+    def __new__(cls, name, bases, attrs):
+        register_cls = super().__new__(cls, name, bases, attrs)
+        # Don't register the base Metric class
+        if name != "Metric":
+            # ensure a single Domain mixin is defined
+            if len(bases) != ALLOWABLE_METRIC_MIXINS + 1 or not any(
+                issubclass(base_type, Domain) for base_type in bases
+            ):
+                raise MixinTypeError(name, "Domain")
+            try:
+                metric_name = attrs["name"]
+            except KeyError:
+                raise MissingAttributeError(name, "name")
+            MetaMetric._registry[metric_name] = register_cls
+        return register_cls
+
+    @classmethod
+    def get_registered_metric_class_from_metric_name(cls, metric_name: str) -> type["Metric"]:
+        """Returns the registered Metric class for a given metric name."""
+        return cls._registry[metric_name]
```

Comment:
This can throw a `KeyError` which might not produce a good error message. Maybe try/catch to produce a better error message, eg: `No metric named metric_name`.